VDA-XP-Reward/contracts/VDAXPReward.sol

```diff
@@ -0,0 +1,277 @@
+        }
+        uint32 orgVal = conversionRate;
+        conversionRate = newRate;
+        emit UpdateConversionRate(orgVal, newRate);
+    }
+
+    /**
+     * @notice Validate the time of proof
+     * @dev Separated for `stack too deep`
+     * @param didAddress - DID address. Used to revert
+     * @param curTime - Current time
+     * @param info - Claim information
+     */
+    function _validateProofTime(address didAddress, DateTime._DateTime memory curTime, ClaimInfo calldata info) internal virtual view {
+        if (curTime.year != info.issueYear || curTime.month != (info.issueMonth + 1)) {
+            revert InvalidProofTime(didAddress, info.typeId, info.issueYear, info.issueMonth);
+        } else if (curTime.month == 1) {
+            if (curTime.year != (info.issueYear + 1) || info.issueMonth != 12) {
+                revert InvalidProofTime(didAddress, info.typeId, info.issueYear, info.issueMonth);
+            }
+        }
+    }
+
+    /**
+     * @notice Validate proof signer
+     * @dev Separated for `stack too deep`
+     * @param didAddress - DID address, Used to revert
+     * @param proofSigner - Signer of proof
+     * @param info - Claim information
+     */
+    function _validateProofSigner(address didAddress, address proofSigner, ClaimInfo calldata info) internal virtual view {
+        bool isVerified;
+        uint index;
+        uint length = _trustedSigners.length();
+
+        while (index < length && !isVerified) {
+            if (proofSigner == _trustedSigners.at(index)) {
+                isVerified = true;
+            }
+            unchecked { ++index; }
+        }
+
+        if (!isVerified) {
+            revert InvalidProof(didAddress, info.typeId, info.xp, info.signature);
+        }
+    }
+
+    /**
+     * @notice Check the proofs and return the total amount of XP
+     * @param didAddress - DID address
+     * @param infos - Array of claim requests
+     * @return uint - Total amount of requested claim
+     */
+    function _validateClaimSignature(address didAddress, ClaimInfo[] calldata infos) internal virtual returns(uint) {
+        uint totalXP;
+
+        bytes memory rawMsg;
+        bytes32 dataHash;
+        address proofSigner;
+
+        uint infoLen = infos.length;
+        DateTime._DateTime memory curTime = DateTime.parseTimestamp(block.timestamp);
+
+        for (uint i; i < infoLen;) {
+            if (infos[i].xp == 0) {
+                revert InvalidXP(didAddress, infos[i].typeId, infos[i].xp);
+            }
+            _validateProofTime(didAddress, curTime, infos[i]);
+
+            rawMsg = abi.encodePacked(didAddress, infos[i].typeId, infos[i].issueYear, infos[i].issueMonth);
+            
+            // Check whether `typeId` is claimed in this month
+            if (isClaimedTypeId[rawMsg]) {
+                revert DuplicatedRequest(didAddress, infos[i].typeId, infos[i].xp, infos[i].signature);
+            }
+            isClaimedTypeId[rawMsg] = true;
```

Comment:
Should this only be set at the very end if all the tests pass?

Reply:
Just not clear yet.
In each for loop, it tests whether the `typeId` is claimed. If claimed, it stops the for loop and revert the error.

Reply:
But, let's say it's a duplicate unique ID (checked later in the code), then it will have already set `isClaimedTypeId` true... however it wasn't actually claimed due to the duplicate unique id.

Reply:
In the solidity,for the reverted transactions, all the status are not updated.
In fact, Solidity is Turing complete. Means that it can know the result before running the transaction. So it doesn't run any code for the reverted transaction.

So, there won't be any cases that `typeId` updated to true for duplicated uniqueId.

Reply:
Ok, got it, thanks.